Source/EasyNetQ/MessageProperties.cs

```diff
@@ -1,362 +1,176 @@
-        if (clusterIdPresent) copy.ClusterId = clusterId;
-
-        if (headers?.Count > 0)
-            copy.headers = new Dictionary<string, object?>(headers);
+    internal MessageProperties(IBasicProperties basicProperties)
+    {
+        ContentType = basicProperties.ContentType;
+        ContentEncoding = basicProperties.ContentEncoding;
+        DeliveryMode = basicProperties.DeliveryMode;
+        Priority = basicProperties.Priority;
+        CorrelationId = basicProperties.CorrelationId;
+        ReplyTo = basicProperties.ReplyTo;
+        Expiration = int.TryParse(basicProperties.Expiration, out var expirationMilliseconds)
+            ? TimeSpan.FromMilliseconds(expirationMilliseconds)
+            : default;
+        MessageId = basicProperties.MessageId;
+        Timestamp = basicProperties.Timestamp.UnixTime;
+        Type = basicProperties.Type;
+        UserId = basicProperties.UserId;
+        AppId = basicProperties.AppId;
+        ClusterId = basicProperties.ClusterId;
+        Headers = basicProperties.Headers switch
+        {
+            null => null,
 
-        return copy;
+            IReadOnlyDictionary<string, object?> readonlyDictionary => readonlyDictionary,
+            _ => new ReadOnlyDictionary<string, object?>(basicProperties.Headers)
+        };
     }
 
-    private bool contentTypePresent;
-    private bool contentEncodingPresent;
-    private bool deliveryModePresent;
-    private bool priorityPresent;
-    private bool correlationIdPresent;
-    private bool replyToPresent;
-    private bool expirationPresent;
-    private bool messageIdPresent;
-    private bool timestampPresent;
-    private bool typePresent;
-    private bool userIdPresent;
-    private bool appIdPresent;
-    private bool clusterIdPresent;
-    private string? contentType;
-
     /// <summary>
     ///     MIME Content type
     /// </summary>
-    public string? ContentType
-    {
-        get => contentType;
-        set
-        {
-            contentType = CheckShortString(value, nameof(ContentType));
-            contentTypePresent = true;
-        }
-    }
-
-    private string? contentEncoding;
+    public string? ContentType { get; init; }
 
     /// <summary>
     ///     MIME content encoding
     /// </summary>
-    public string? ContentEncoding
-    {
-        get => contentEncoding;
-        set
-        {
-            contentEncoding = CheckShortString(value, nameof(ContentEncoding));
-            contentEncodingPresent = true;
-        }
-    }
-
-    private IDictionary<string, object?>? headers;
+    public string? ContentEncoding { get; init; }
 
     /// <summary>
     ///     Various headers
     /// </summary>
-    public IDictionary<string, object?> Headers
-    {
-        get => headers ??= new Dictionary<string, object?>();
-        set => headers = value;
-    }
-
-    private byte deliveryMode;
+    public IReadOnlyDictionary<string, object?>? Headers { get; init; }
```

Comment:
Headers property is not required to be of readonly/immutable type. What is the reason to do so?